File: _site/solutions/146_lru_cache.rb
class LRUCache
  #     :type capacity: Integer
  def initialize(capacity)
    @capacity = capacity
    @cache = {}
  end

  #     :type key: Integer
  #     :rtype: Integer
  def get(key)
    return -1 unless @cache.key?(key)
    val = @cache[key]
    @cache.delete(key)
    @cache[key] = val
    val
  end

  #     :type key: Integer
  #     :type value: Integer
  #     :rtype: Void
  def put(key, value)
    return if @capacity.zero?
    @cache.delete(key)
    @cache.shift while @cache.size >= @capacity
    @cache[key] = value
  end
end

# Your LFUCache object will be instantiated and called as such:
obj = LRUCache.new(2)
obj.put(1, 1)
obj.put(2, 2)
p obj.get(1)
obj.put(3, 3)
p obj.get(2)
obj.put(4, 4)
p obj.get(1)
p obj.get(3)
p obj.get(4)

File: python_solutions/146.lru-cache.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import string
true = True
false = False
MIN, MAX = -0x3f3f3f3f, 0x3f3f3f3f
#
# @lc app=leetcode id=146 lang=python3
#
# [146] LRU Cache
#
# https://leetcode.com/problems/lru-cache/description/
#
# algorithms
# Medium (30.71%)
# Total Accepted:    476.4K
# Total Submissions: 1.5M
# Testcase Example:  '["LRUCache","put","put","get","put","get","put","get","get","get"]\n' +
# '[[2],[1,1],[2,2],[1],[3,3],[2],[4,4],[1],[3],[4]]'
#
# Design and implement a data structure for Least Recently Used (LRU) cache. It
# should support the following operations: get and put.
#
# get(key) - Get the value (will always be positive) of the key if the key
# exists in the cache, otherwise return -1.
# put(key, value) - Set or insert the value if the key is not already present.
# When the cache reached its capacity, it should invalidate the least recently
# used item before inserting a new item.
#
# The cache is initialized with a positive capacity.
#
# Follow up:
# Could you do both operations in O(1) time complexity?
#
# Example:
#
#
# LRUCache cache = new LRUCache( 2 /* capacity */ );
#
# cache.put(1, 1);
# cache.put(2, 2);
# cache.get(1);       // returns 1
# cache.put(3, 3);    // evicts key 2
# cache.get(2);       // returns -1 (not found)
# cache.put(4, 4);    // evicts key 1
# cache.get(1);       // returns -1 (not found)
# cache.get(3);       // returns 3
# cache.get(4);       // returns 4
#
#
#
#
#


class LRUCache:

    def __init__(self, capacity: int):
        self.ds = OrderedDict()
        self.size = capacity

    def get(self, key: int) -> int:
        if key not in self.ds:
            return -1
        v = self.ds.pop(key)
        self.ds[key] = v
        return v

    def put(self, key: int, value: int) -> None:
        if key in self.ds:
            self.ds.pop(key)
        else:
            if self.size > 0:
                self.size -= 1
            else:
                self.ds.popitem(last=False)

        self.ds[key] = value


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

# sol = Solution()
